File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/LiveOverviewFrame.kt
package com.phren.kit

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/** One frame of the Hook's `/v1/overview` stream: the overview when it changed, or a heartbeat saying the phone's copy is current. */
sealed interface LiveOverviewFrame {
    data class Overview(val workspaces: LiveWorkspaces) : LiveOverviewFrame
    data class Heartbeat(val info: LiveHookInfo?) : LiveOverviewFrame

    companion object {
        /** Heartbeats are small; an overview is read with the same checks as a polled `/v1/workspaces` answer, Hook envelope included. */
        fun read(data: ByteArray): LiveOverviewFrame {
            if (data.size <= 16_384) {
                val envelope = runCatching { hookJson.parseToJsonElement(data.decodeToString()) as? JsonObject }.getOrNull()
                    ?: throw PhrenKitError.Validation("Phren Hook sent an unknown overview frame.")
                when (envelope["type"].str) {
                    "heartbeat" -> {
                        // The heartbeat's `phren` block has the overview envelope's shape.
                        val info = envelope["phren"].obj?.let { phren ->
                            LiveWorkspaces.read(buildJsonObject { put("kind", JsonPrimitive("herdr")); put("groups", kotlinx.serialization.json.JsonArray(emptyList())); put("phren", phren) }).phren
                        }
                        return Heartbeat(info)
                    }
                    "overview" -> Unit
                    else -> throw PhrenKitError.Validation("Phren Hook sent an unknown overview frame.")
                }
            }
            return Overview(LiveWorkspaces.read(data, requiringHook = true))
        }
    }
}

File: apps/ios/PhrenKit/Sources/PhrenKit/Sessions/LiveOverviewFrame.swift
import Foundation

/// One frame of the Hook's `/v1/overview` stream: the overview when it
/// changed, or a heartbeat saying the overview the phone holds is current.
public enum LiveOverviewFrame: Equatable, Sendable {
    case overview(LiveWorkspaces)
    case heartbeat(LiveHookInfo?)

    private struct Envelope: Decodable {
        let type: String
        let phren: LiveHookInfo?
    }

    /// Heartbeats are small; an overview is read with the same checks as a
    /// polled `/v1/workspaces` answer, including the Hook envelope.
    public static func read(_ data: Data) throws -> Self {
        if data.count <= 16_384 {
            let envelope = try JSONDecoder().decode(Envelope.self, from: data)
            switch envelope.type {
            case "heartbeat": return .heartbeat(envelope.phren)
            case "overview": break
            default: throw PhrenKitError.validation("Phren Hook sent an unknown overview frame.")
            }
        }
        return .overview(try LiveWorkspaces.read(data, requiringHook: true))
    }
}
